### DISCUSSION TITLE
Documentation on how to retrieve the from, to, subject and body

### DISCUSSION THREAD

**Discussion by kishaningithub** [bebleo/smtpdfix#333] [Q&A]
The test code in  [README documentation](https://github.com/bebleo/smtpdfix#using) asserts for messages count but i would like to know how to get the from, to, subject and body from the email that was sent

```python
from smtplib import SMTP

def test_send_mail(smtpd):
    with SMTP(smtpd.hostname, smtpd.port) as client:
        send_email() # This is the function under test

    assert len(smtpd.messages) == 1
    assert smtpd.messages[0]. # Problem - I cant seem to find the attributes to retrieve from, to, subject and body
```

Can you help?

**Comment by bebleo** [comment]
Accessing the value of a header for an email object in python is relatively easy provided you know the key, but accessing the body is more difficult.

Keys for email headers are case-agnostic (e.g., "to", "To", and "TO" are equally acceptable). It may, therefore, be necessary to determine the key before you can retrieve the value. The example provided below does not account for this possibility nor for many other cases.

To retrieve the body it is important to note that `smtpd.messages` is a list of `email.message.Message` objects received by the SMTP server, SMTPDFix. The following example is a naive example of creating an email, sending it using `smtplib.SMTP` to a SMTPDFix server, and then retrieving the headers and body of the received email.

```python
def test_email_headers_and_body(self, smtpd: AuthController) -> NoReturn:
    '''
    Get email headers and body once received.
    '''

    def _get_body(
            message: Union[Message, EmailMessage]
            ) -> Union[bytes, None, str]:
        '''
        Helper method to extract the body from an email message.
        Code not tested in production.
        '''

        if message.is_multipart():
            # handle a multipart message
            for part in message.walk():
                dispo = part.get_content_disposition()
                if (part.get_content_type() == "text/plain" and
                        dispo != "attachment"):
                    return part.get_payload(decode=True)

        return message.get_payload(decode=True)

    # Generate an email.message.EmailMessage to send
    sender = "from@example.org"
    to = ["recipient1@example.org", "recipient2@example.org"]
    subject = "This is an email"
    content = "This is plain text email body"

    msg = EmailMessage()
    msg["Sender"] = sender
    msg["Reply-To"] = sender
    msg["To"] = ", ".join(to)
    msg["Subject"] = subject
    msg.set_content(content)

    # Send the message using smtplib.SMTP
    with SMTP(smtpd.hostname, smtpd.port) as client:
        client.send_message(msg)

    # Ensure we have received only one message!
    assert len(smtpd.messages) == 1

    # Test that the message is as expected.
    for rcvd in smtpd.messages:
        assert rcvd["Sender"] == msg["Sender"]
        assert rcvd["Reply-To"] == sender
        body = _get_body(rcvd)
        if isinstance(body, bytes):
            body = body.decode()
        assert content in body
```

A more complete answer to the question, for anyone stumbling across this discussion in the future, must include that it may make more sense to write tests against the `send_email()` function. Those tests could ensure that the generated email includes the expected content, is calling a function to transmit the email with the expected values, etc.

While I am delighted when I see SMTPFix being used, it is intended to help fill a need for cases that require a SMTP server for testing. There may be, and in my opinion often are, better ways to test email related code.

Note: While out of scope for this discussion it may be useful to note that `email.message.Message` has been replaced by `email.message.EmailMessage` as the suggested high-level object for handling email messages as of python 3.6. I plan to have a future version of SMTPDFix and pytest-smtpd return `smtpd.messages` as a list of `email.message.EmailMessage`; as of mid-July 2023, however, there is no timeline for this change to be implemented.
